### PR TITLE
Replace Flask by Sanic

### PR DESCRIPTION
**Pull request by cowjen01**
**Proposed changes**:
- replace Flask by sanic ( #109 )
- make actions executor asynchronous
- improve endpoints tests
- improve type checks
- add sanic workers number available

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa-sdk#code-style) for instructions)
- [x] updated the changelog
- [ ] updated the documentation

### CODE REVIEW

**Comment by erohmensing** [comment]
Thanks for the PR! We'll give it a review as soon as possible.

**Comment by tmbo** [comment]
@wochinge would be great if you can review this. I had a quick look but won't have enough time to look at this in detail anytime soon. 

Things we need to check:
- ssl / cors handling
- I like the async `run` method of actions, but we need to be able to still support non async methods (meaning classes implementing a normal `def run` as well as functions directly getting registered)
- the above should also make sure that this is backwards compatible

**Comment by cowjen01** [comment]
@tmbo thanks for the quick review. Meanwhile you are taking a detailed look on our changes, we can update the executor so it will support both sync and async run functions.

File: setup.py [hunk 1]
```diff
@@ -16,12 +16,10 @@
...
 
 install_requires = [
-    "requests>=2.20",  # for common use in custom actions code
     "ConfigArgParse>=0.14",
     "coloredlogs~=10.0",
-    "flask~=1.0",
-    "flask_cors~=3.0",
-    "gevent>=1.4",
+    "sanic~=19.6",
+    "sanic-cors==0.9.9.post1",
```
Comment (wochinge) [inline setup.py]: how about going straight to `0.9.9.post3`?

**Comment by wochinge** [comment]
Saw that Tom already mentioned the compatibility issues 🙈

File: rasa_sdk/interfaces.py [hunk 2]
```diff
@@ -196,3 +196,14 @@ def __init__(self, action_name: Text, message: Optional[Text] = None) -> None:
 
     def __str__(self) -> Text:
         return self.message
+
+
+class ActionNotFoundRejection(Exception):
```
[the remark this answers is not in the document]
Comment (cowjen01) [reply rasa_sdk/interfaces.py]: absolutely! will do

File: rasa_sdk/utils.py [hunk 3]
```diff
@@ -76,3 +83,82 @@ def arguments_of(func) -> List[Text]:
     """Return the parameters of the function `func` as a list of their names."""
 
     return inspect.signature(func).parameters.keys()
+
+
+def number_of_sanic_workers() -> int:
```
[the remark this answers is not in the document]
Comment (cowjen01) [reply rasa_sdk/utils.py]: yes, no problem

File: rasa_sdk/utils.py [hunk 4]
```diff
@@ -76,3 +83,82 @@ def arguments_of(func) -> List[Text]:
...
+            f"Cannot convert environment variable `{ENV_SANIC_WORKERS}` "
+            f"to int ('{os.environ[ENV_SANIC_WORKERS]}')."
+        )
+        return _log_and_get_default_number_of_workers()
+
+    if env_value == DEFAULT_SANIC_WORKERS:
+        return _log_and_get_default_number_of_workers()
+
+    if env_value < 1:
+        logger.debug(
```
[the remark this answers is not in the document]
Comment (cowjen01) [reply rasa_sdk/utils.py]: np, copy that

File: requirements.txt [hunk 5]
```diff
@@ -1,7 +1,5 @@
 ## Setup py requirements
-requests==2.22.0
 ConfigArgParse==0.15.1
 coloredlogs==10.0
-flask==1.1.1
-flask_cors==3.0.8
-gevent==1.4.0
+sanic==19.6.3
```
[the remark this answers is not in the document]
Comment (cowjen01) [reply requirements.txt]: I thought it should copy the version from rasa package, but I can update to 19.9.0 if it's ok for you

File: setup.py (hunk 1 above)
Comment (cowjen01) [reply setup.py]: same as the sanic package

File: rasa_sdk/endpoint.py [hunk 6]
```diff
@@ -39,120 +55,73 @@ def create_argument_parser():
...
+        return create_ok_response(result)
 
-    @app.route("/actions", methods=["GET", "OPTIONS"])
-    @cross_origin(origins=cors_origins)
-    def actions():
+    @app.get("/actions")
+    async def actions(request):
         """List all registered actions."""
-        return jsonify([{"name": k} for k in executor.actions.keys()])
+        return create_ok_response([{"name": k} for k in executor.actions.keys()])
```
[the remark this answers is not in the document]
Comment (cowjen01) [reply rasa_sdk/endpoint.py]: no problem

File: rasa_sdk/endpoint.py [hunk 7]
```diff
@@ -39,120 +55,73 @@ def create_argument_parser():
...
+def create_error_response(
+    message: Text, action_name: Text, status: int = 500
+) -> HTTPResponse:
+    response = json({"error": message, "action_name": action_name})
+    response.status = status
+    return response
+
+
+def create_app(
+    action_package_name: Union[Text, types.ModuleType],
```
[the remark this answers is not in the document]
Comment (cowjen01) [reply rasa_sdk/endpoint.py]: thank you! appreciate that

File: rasa_sdk/endpoint.py [hunk 8]
```diff
@@ -39,120 +55,73 @@ def create_argument_parser():
     return parser
 
 
-def endpoint_app(cors_origins=None, action_package_name=None):
-    app = Flask(__name__)
+def create_ok_response(body: Any, status: int = 200) -> HTTPResponse:
```
[the remark this answers is not in the document]
Comment (cowjen01) [reply rasa_sdk/endpoint.py]: roger

File: rasa_sdk/endpoint.py [hunk 9]
```diff
@@ -39,120 +55,73 @@ def create_argument_parser():
...
-    def webhook():
+    @app.post("/webhook")
+    async def webhook(request):
         """Webhook to retrieve action calls."""
         action_call = request.json
-        check_version_compatibility(action_call.get("version"))
+        utils.check_version_compatibility(action_call.get("version"))
         try:
-            response = executor.run(action_call)
+            result = await executor.run(action_call)
```
[the remark this answers is not in the document]
Comment (cowjen01) [reply rasa_sdk/endpoint.py]: Yes, this is a compatibility problem and it definitely should support both approaches

File: rasa_sdk/endpoint.py [hunk 10]
```diff
@@ -39,120 +55,73 @@ def create_argument_parser():
...
-    app = Flask(__name__)
+def create_ok_response(body: Any, status: int = 200) -> HTTPResponse:
+    response = json(body)
+    response.status = status
+    return response
+
 
-    if not cors_origins:
-        cors_origins = []
+def create_error_response(
```
[the remark this answers is not in the document]
Comment (cowjen01) [reply rasa_sdk/endpoint.py]: unfortunately there is no method `error` of `response` object - is this some custom implementation? for now I let it like it was before my changes:

```python
body = {"error": e.message, "action_name": e.action_name}
return response.json(body, status=404)
```

File: rasa_sdk/utils.py [hunk 11]
```diff
@@ -76,3 +83,82 @@ def arguments_of(func) -> List[Text]:
...
+            f"Cannot convert environment variable `{ENV_SANIC_WORKERS}` "
+            f"to int ('{os.environ[ENV_SANIC_WORKERS]}')."
+        )
+        return _log_and_get_default_number_of_workers()
+
+    if env_value == DEFAULT_SANIC_WORKERS:
+        return _log_and_get_default_number_of_workers()
+
+    if env_value < 1:
+        logger.warning(
```
Comment (wochinge) [inline rasa_sdk/utils.py]: Let's use `warnings` because it's something the user can and should fix (https://stackoverflow.com/a/14762106/3429596)

```suggestion
        warnings.warn(
```

File: rasa_sdk/utils.py [hunk 12]
```diff
@@ -76,3 +83,82 @@ def arguments_of(func) -> List[Text]:
...
+            f"Cannot set number of Sanic workers to the desired value "
+            f"({env_value}). The number of workers must be at least 1."
+        )
+        return _log_and_get_default_number_of_workers()
+
+    logger.debug(f"Using {env_value} Sanic workers.")
+    return env_value
+
+
+def check_version_compatibility(rasa_version):
```
Comment (wochinge) [inline rasa_sdk/utils.py]:
```suggestion
def check_version_compatibility(rasa_version: Optional[Text]) -> None:
```

File: tests/test_endpoint.py [hunk 13]
```diff
@@ -1,7 +1,47 @@
...
+    assert response.status == 404
+
+
+def test_server_webhook_custom_action_returns_200():
+    data = {
+        "next_action": "custom_action",
+        "tracker": {"sender_id": "1", "conversation_id": "default"},
+    }
+    request, response = app.test_client.post("/webhook", data=json.dumps(data))
+    assert response.status == 200
```
Comment (wochinge) [inline tests/test_endpoint.py]: can we please also `assert` the response content?

File: tests/test_endpoint.py [hunk 14]
```diff
@@ -1,7 +1,47 @@
...
+
+
+def test_server_webhook_custom_async_action_returns_200():
+    data = {
+        "next_action": "custom_async_action",
+        "tracker": {"sender_id": "1", "conversation_id": "default"},
+    }
+    request, response = app.test_client.post("/webhook", data=json.dumps(data))
```
Comment (wochinge) [inline tests/test_endpoint.py]: can we please also `assert` the response content?

File: tests/test_endpoint.py (hunk 13 above)
Comment (wochinge) [reply tests/test_endpoint.py]: maybe we could even add an event in the test actions

File: tests/test_utils.py [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+from rasa_sdk.utils import number_of_sanic_workers
+from rasa_sdk.constants import DEFAULT_SANIC_WORKERS, ENV_SANIC_WORKERS
+
+
+def test_default_number_of_sanic_workers():
+    n = number_of_sanic_workers()
+    assert n == DEFAULT_SANIC_WORKERS
+
+
+def test_env_number_of_sanic_workers():
```
Comment (wochinge) [inline tests/test_utils.py]: optional suggestion: use `parametrize` (https://docs.pytest.org/en/latest/parametrize.html)

File: tests/test_actions.py [hunk 16]
```diff
@@ -3,6 +3,15 @@
 from rasa_sdk.executor import ActionExecutor, CollectingDispatcher
 
 
+class CustomAsyncAction(Action):
```
Comment (wochinge) [inline tests/test_actions.py]: can we please add types to this one, too?

File: rasa_sdk/executor.py [hunk 17]
```diff
@@ -229,24 +229,26 @@ def validate_events(events: List[Dict[Text, Any]], action_name: Text):
...
-                )
+                raise ActionNotFoundException(action_name)
 
             tracker_json = action_call.get("tracker")
             domain = action_call.get("domain", {})
             tracker = Tracker.from_dict(tracker_json)
             dispatcher = CollectingDispatcher()
 
-            events = action(dispatcher, tracker, domain)
+            if inspect.iscoroutinefunction(action):
```
Comment (wochinge) [inline rasa_sdk/executor.py]: 🚀 Optional suggestion: Extract it in a function and add some tests

File: rasa_sdk/endpoint.py [hunk 18]
```diff
@@ -39,120 +54,63 @@ def create_argument_parser():
...
 
-    @app.route("/webhook", methods=["POST", "OPTIONS"])
-    @cross_origin()
-    def webhook():
+    @app.post("/webhook")
+    async def webhook(request):
         """Webhook to retrieve action calls."""
         action_call = request.json
-        check_version_compatibility(action_call.get("version"))
+        utils.check_version_compatibility(action_call.get("version"))
```
Comment (wochinge) [inline rasa_sdk/endpoint.py]: can we please also handle the case that `action_call` is `None` (sorry, I know it's not your code, but it should be a quick fix)

File: rasa_sdk/utils.py [hunk 19]
```diff
@@ -76,3 +83,82 @@ def arguments_of(func) -> List[Text]:
...
+            "both rasa and rasa_sdk."
+            "".format(rasa_sdk.__version__)
+        )
+        return
+
+    rasa = rasa_version.split(".")[:-1]
+    sdk = rasa_sdk.__version__.split(".")[:-1]
+
+    if rasa != sdk:
+        logger.warning(
```
Comment (wochinge) [inline rasa_sdk/utils.py]:
```suggestion
        warnings.warn(
```

File: rasa_sdk/utils.py [hunk 20]
```diff
@@ -76,3 +83,82 @@ def arguments_of(func) -> List[Text]:
...
+    Args:
+        rasa_version - A string containing the version of rasa that
+        is making the call to the action server.
+    Raises:
+        Warning - The version of rasa version unknown or not compatible with
+        this version of rasa_sdk.
+    """
+    # Check for versions of Rasa that are too old to report their version number
+    if rasa_version is None:
+        logger.warning(
```
Comment (wochinge) [inline rasa_sdk/utils.py]:
```suggestion
        warnings.warn(
```

File: setup.py [hunk 21]
```diff
@@ -16,12 +16,10 @@
...
 
 install_requires = [
-    "requests>=2.20",  # for common use in custom actions code
     "ConfigArgParse>=0.14",
     "coloredlogs~=10.0",
-    "flask~=1.0",
-    "flask_cors~=3.0",
-    "gevent>=1.4",
+    "sanic~=19.9",
+    "sanic-cors==0.9.9.post3",
```
Comment (wochinge) [inline setup.py]: Sorry, I made a bad suggestion. `rasa` currently is only compatible with `0.9.9.post1` so we should also have `0.9.9.post1` here to avoid version conflicts
```suggestion
    "sanic-cors==0.9.9.post1",
```

File: requirements.txt [hunk 22]
```diff
@@ -1,7 +1,5 @@
 ## Setup py requirements
-requests==2.22.0
 ConfigArgParse==0.15.1
 coloredlogs==10.0
-flask==1.1.1
-flask_cors==3.0.8
-gevent==1.4.0
+sanic==19.9.0
+sanic-cors==0.9.9.post3
```
Comment (wochinge) [inline requirements.txt]: See my comment below
```suggestion
sanic-cors==0.9.9.post1
```

File: rasa_sdk/endpoint.py [hunk 23]
```diff
@@ -39,120 +54,63 @@ def create_argument_parser():
...
     executor = ActionExecutor()
     executor.register_package(action_package_name)
 
-    CORS(app, resources={r"/*": {"origins": cors_origins}})
-
-    @app.route("/health", methods=["GET", "OPTIONS"])
-    @cross_origin(origins=cors_origins)
-    def health():
+    @app.get("/health")
+    async def health(request):
```
Comment (wochinge) [inline rasa_sdk/endpoint.py]:
```suggestion
    async def health(_) -> HTTPResponse:
```

File: rasa_sdk/endpoint.py [hunk 24]
```diff
@@ -39,120 +54,63 @@ def create_argument_parser():
...
         """Ping endpoint to check if the server is running and well."""
-        return jsonify({"status": "ok"})
+        body = {"status": "ok"}
+        return response.json(body, status=200)
 
-    @app.route("/webhook", methods=["POST", "OPTIONS"])
-    @cross_origin()
-    def webhook():
+    @app.post("/webhook")
+    async def webhook(request):
```
Comment (wochinge) [inline rasa_sdk/endpoint.py]:
```suggestion
    async def webhook(request: Request) -> HTTPResponse:
```

File: rasa_sdk/endpoint.py [hunk 25]
```diff
@@ -39,120 +54,63 @@ def create_argument_parser():
...
+            return response.json(body, status=400)
+        except ActionNotFoundException as e:
+            logger.error(e)
+            body = {"error": e.message, "action_name": e.action_name}
+            return response.json(body, status=404)
+
+        return response.json(result, status=200)
+
+    @app.get("/actions")
+    async def actions(request):
```
Comment (wochinge) [inline rasa_sdk/endpoint.py]:
```suggestion
    async def actions(_) -> HTTPResponse:
```

File: rasa_sdk/endpoint.py [hunk 26]
```diff
@@ -39,120 +54,63 @@ def create_argument_parser():
...
-
 def run(
-    action_package_name,
-    port=DEFAULT_SERVER_PORT,
-    cors_origins="*",
-    ssl_certificate=None,
-    ssl_keyfile=None,
-    ssl_password=None,
+    action_package_name: Union[Text, types.ModuleType],
+    port: int = DEFAULT_SERVER_PORT,
```
Comment (wochinge) [inline rasa_sdk/endpoint.py]: I think this can also be a text (doesn't needs to be handled, I tried)
```suggestion
    port: Union[Text, int] = DEFAULT_SERVER_PORT,
```

**Comment by wochinge** [comment]
and please update the changelog :-)

File: Makefile [hunk 27]
```diff
@@ -32,7 +32,7 @@ lint:
 	black --check rasa_sdk tests
 
 test: clean
-	py.test tests --cov rasa_sdk
+	py.test tests --cov rasa_sdk -W ignore
```
Comment (wochinge) [inline Makefile]: why this change?

File: rasa_sdk/executor.py [hunk 28]
```diff
@@ -6,7 +6,9 @@
 from typing import Text, List, Dict, Any, Type, Union, Callable, Optional
 import typing
 import types
-from rasa_sdk.interfaces import Tracker
+
+from rasa_sdk.interfaces import Tracker, ActionNotFoundException
+from rasa_sdk.utils import is_coroutine_action
```
Comment (wochinge) [inline rasa_sdk/executor.py]:
```suggestion
```

File: rasa_sdk/executor.py [hunk 29]
```diff
@@ -254,24 +256,26 @@ def validate_events(events: List[Dict[Text, Any]], action_name: Text):
...
-                )
+                raise ActionNotFoundException(action_name)
 
             tracker_json = action_call.get("tracker")
             domain = action_call.get("domain", {})
             tracker = Tracker.from_dict(tracker_json)
             dispatcher = CollectingDispatcher()
 
-            events = action(dispatcher, tracker, domain)
+            if is_coroutine_action(action):
```
Comment (wochinge) [inline rasa_sdk/executor.py]:
```suggestion
            if utils.is_coroutine_action(action):
```

File: CHANGELOG.rst [hunk 30]
```diff
@@ -13,9 +13,17 @@ This project adheres to `Semantic Versioning`_ starting with version 0.11.0.
 
 Added
 -----
+- add unit tests of endpoints, actions and utils
```
Comment (wochinge) [inline CHANGELOG.rst]: We don't need a changelog update for the tests
```suggestion
```

File: CHANGELOG.rst [hunk 31]
```diff
@@ -13,9 +13,17 @@ This project adheres to `Semantic Versioning`_ starting with version 0.11.0.
...
+- add support of multiple sanic workers
```
Comment (wochinge) [inline CHANGELOG.rst]:
```suggestion
- Add support for multiple sanic workers (configurable with the ``ACTION_SERVER_SANIC_WORKERS` environment variable)
```

File: CHANGELOG.rst [hunk 32]
```diff
@@ -13,9 +13,17 @@ This project adheres to `Semantic Versioning`_ starting with version 0.11.0.
...
+- add support of async run method inside the action class
```
Comment (wochinge) [inline CHANGELOG.rst]:
```suggestion
- Add support for async ``run`` methods in the ``Action`` class
```

File: CHANGELOG.rst [hunk 33]
```diff
@@ -13,9 +13,17 @@ This project adheres to `Semantic Versioning`_ starting with version 0.11.0.
...
+- add typing for methods across the package
```
Comment (wochinge) [inline CHANGELOG.rst]: Same as for the tests (We appreciate both changes a lot 🚀 )
```suggestion
```

File: CHANGELOG.rst [hunk 34]
```diff
@@ -13,9 +13,17 @@ This project adheres to `Semantic Versioning`_ starting with version 0.11.0.
...
+- add handler of unknown action
+- add handler of empty body requests sent to webhook endpoint
```
Comment (wochinge) [inline CHANGELOG.rst]: Please also move this to ``Changed``
```suggestion
- Return status code ``400`` in case an empty request body was sent to the ``/webhook`` endpoint
```

File: CHANGELOG.rst [hunk 35]
```diff
@@ -13,9 +13,17 @@ This project adheres to `Semantic Versioning`_ starting with version 0.11.0.
...
 
 Changed
 -------
+- replace ``flask`` server framework by ``sanic``
```
Comment (wochinge) [inline CHANGELOG.rst]:
```suggestion
- Replace ``flask`` server framework with ``sanic``
```

File: CHANGELOG.rst [hunk 36]
```diff
@@ -13,9 +13,17 @@ This project adheres to `Semantic Versioning`_ starting with version 0.11.0.
...
+- replace ``flask_cors`` by ``sanic-cors``
```
Comment (wochinge) [inline CHANGELOG.rst]:
```suggestion
- Replace ``flask_cors`` with ``sanic-cors``
```

File: CHANGELOG.rst [hunk 37]
```diff
@@ -28,6 +36,9 @@ Fixed
 
 Removed
 -------
+- remove ``flask`` framework and ``flask_cors``
```
Comment (wochinge) [inline CHANGELOG.rst]: I think that's already covered by the changelog above
```suggestion
```

File: CHANGELOG.rst [hunk 38]
```diff
@@ -28,6 +36,9 @@ Fixed
...
+- remove ``requests`` package
```
Comment (wochinge) [inline CHANGELOG.rst]:
```suggestion
- Remove ``requests`` dependency
```

File: CHANGELOG.rst [hunk 39]
```diff
@@ -28,6 +36,9 @@ Fixed
...
+- remove ``gevent`` package
```
Comment (wochinge) [inline CHANGELOG.rst]:
```suggestion
- Remove ``gevent`` dependency
```

File: CHANGELOG.rst [hunk 40]
```diff
@@ -13,9 +13,17 @@ This project adheres to `Semantic Versioning`_ starting with version 0.11.0.
 
 Added
 -----
+- add unit tests of endpoints, actions and utils
+- add support of multiple sanic workers
+- add support of async run method inside the action class
+- add typing for methods across the package
+- add handler of unknown action
```
Comment (wochinge) [inline CHANGELOG.rst]: Please also move this to `Changed` :-) 
```suggestion
- Return status code ``404` in case requested action was not found
```

File: Makefile (hunk 27 above)
Comment (cowjen01) [reply Makefile]: when we've changed `logger.warning` to `warnings.warn` it has started to print warnings about invalid number of sanic workers or rasa version while the tests were running and because of that it never fully passed

**Comment by cowjen01** [comment]
@wochinge Thank you for all your recommendations and feedbacks about the code and changelog! 🙂It was my very first PR to Rasa repository and your guidance really helped to get the things done 👍I'm looking forward to contribute in the future again

File: CHANGELOG.rst [hunk 41]
```diff
@@ -13,9 +13,15 @@ This project adheres to `Semantic Versioning`_ starting with version 0.11.0.
 
 Added
 -----
+- Add support for multiple sanic workers (configurable with the ``ACTION_SERVER_SANIC_WORKERS` environment variable)
```
Comment (wochinge) [inline CHANGELOG.rst]:
```suggestion
- Add support for multiple sanic workers (configurable with the ``ACTION_SERVER_SANIC_WORKERS`` environment variable)
```

File: CHANGELOG.rst [hunk 42]
```diff
@@ -13,9 +13,15 @@ This project adheres to `Semantic Versioning`_ starting with version 0.11.0.
...
+- Add support for async ``run`` methods in the ``Action`` class
+- Return status code ``404` in case requested action was not found
```
Comment (wochinge) [inline CHANGELOG.rst]:
```suggestion
- Return status code ``404`` in case requested action was not found
```

File: CHANGELOG.rst (hunk 41 above)
Comment (wochinge) [reply CHANGELOG.rst]: Sorry, seems I missed some back ticks in the suggestion 🙈

Comment (wochinge) [reply CHANGELOG.rst]: I'll merge and add them quickly myself :-)

**Comment by erohmensing** [comment]
Awesome work @cowjen01! 🎉

File: CHANGELOG.rst (hunk 41 above)
Comment (wochinge) [reply CHANGELOG.rst]: https://github.com/RasaHQ/rasa-sdk/pull/116

File: Makefile (hunk 27 above)
Comment (erohmensing) [reply Makefile]: @imLew let's not forget about sdk eventually when fixing the warnings in tests

Comment (imLew) [reply Makefile]: thanks for the reminder ;)

**Comment by fanlu** [comment]
FormAction does not work after upgrade rasa_sdk version to 1.5.0
this is error msg from rasax console:
```
2019-11-27 15:46:08 ERROR    rasa.core.processor  - Encountered an exception while running action 'delivery_form'. Bot will continue, but the actions events are lost. Please check the logs of your action server for more information.
```
this is from action server with rasa_sdk 1.5.0:
```
2019-11-27 15:46:08 DEBUG    rasa_sdk.executor  - Finished running 'action_fetch_profile'
2019-11-27 15:46:08 DEBUG    sanic.root  - CORS: Request to '/webhook' matches CORS resource '/*'. Using options: {'origins': [''], 'methods': 'DELETE, GET, HEAD, OPTIONS, PATCH, POST, PUT', 'allow_headers': ['.*'], 'expose_headers': None, 'supports_credentials': False, 'max_age': None, 'send_wildcard': False, 'automatic_options': True, 'vary_header': True, 'resources': {'/*': {'origins': ''}}, 'intercept_exceptions': True, 'always_send': True}
2019-11-27 15:46:08 INFO     sanic.access  - 
2019-11-27 15:46:08 DEBUG    rasa_sdk.executor  - Received request to run 'delivery_form'
2019-11-27 15:46:08 DEBUG    actions  - There is no active form
2019-11-27 15:46:08 DEBUG    actions  - Activated the form 'delivery_form'
2019-11-27 15:46:08 DEBUG    actions  - No pre-filled required slots to validate.
2019-11-27 15:46:08 DEBUG    actions  - Skipping validation
2019-11-27 15:46:08 DEBUG    actions  - Request next slot 'customerNameConfirm'
2019-11-27 15:46:08 DEBUG    rasa_sdk.executor  - Finished running 'delivery_form'
2019-11-27 15:46:08 DEBUG    sanic.root  - CORS: Request to '/webhook' matches CORS resource '/*'. Using options: {'origins': [''], 'methods': 'DELETE, GET, HEAD, OPTIONS, PATCH, POST, PUT', 'allow_headers': ['.*'], 'expose_headers': None, 'supports_credentials': False, 'max_age': None, 'send_wildcard': False, 'automatic_options': True, 'vary_header': True, 'resources': {'/*': {'origins': ''}}, 'intercept_exceptions': True, 'always_send': True}
2019-11-27 15:46:08 INFO     sanic.access  - 
```

**Comment by wochinge** [comment]
hi @fanlu, thanks for your feedback. I'll check it out right away!

**Comment by wochinge** [comment]
@fanlu fixed in rasa 1.5.1 and this pr https://github.com/RasaHQ/rasa/pull/4855

**Comment by fanlu** [comment]
cool!
